workers/unity/Packages/com.improbable.gdk.core/Worker/ConnectionUtils.cs

```diff
@@ -0,0 +1,135 @@
+        }
+
+        /// <summary>
+        ///     Connects to the SpatialOS Runtime via the Alpha Locator service and creates a <see cref="Connection" /> object
+        ///     asynchronously.
+        /// </summary>
+        /// <param name="parameters">
+        ///     The <see cref="AlphaLocatorConfig" /> object stores the configuration needed to connect via the
+        ///     Locator Service.
+        /// </param>
+        /// <param name="connectionParameters">The <see cref="ConnectionParameters" /> storing </param>
+        /// <returns>
+        ///     A <see cref="Task{TResult}" /> to run this method asynchronously and retrieve the created
+        ///     <see cref="Connection" /> object upon connecting successfully.
+        /// </returns>
+        public static async Task<Connection> CreateConnectionAsync(
+            AlphaLocatorConfig parameters,
+            ConnectionParameters connectionParameters)
+        {
+            using (var locator = new AlphaLocator(parameters.LocatorHost, parameters.LocatorParameters))
+            {
+                using (var connectionFuture = locator.ConnectAsync(connectionParameters))
+                {
+                    return await TryToConnectAsync(connectionFuture);
+                }
+            }
+        }
+
+        private static async Task<Connection> TryToConnectAsync(Future<Connection> connectionFuture)
+        {
+            var connection = await Task.Run(() => connectionFuture.Get()).ConfigureAwait(false);
+            if (connection.GetConnectionStatusCode() != ConnectionStatusCode.Success)
+            {
+                throw new ConnectionFailedException(GetConnectionFailureReason(connection),
+                    ConnectionErrorReason.CannotEstablishConnection);
+            }
+
+            return connection;
+        }
+
+
+        private static async Task<DeploymentList> GetDeploymentList(Locator locator)
+        {
+            using (var deploymentsFuture = locator.GetDeploymentListAsync())
+            {
+                var deploymentList = await Task.Run(() => deploymentsFuture.Get()).ConfigureAwait(false);
+                // Guard against null refs. This shouldn't be triggered.
+                if (!deploymentList.HasValue)
+                {
+                    throw new ConnectionFailedException("Deployment list future returned null.",
+                        ConnectionErrorReason.DeploymentNotFound);
+                }
+
+                if (deploymentList.Value.Error != null)
+                {
+                    throw new ConnectionFailedException(deploymentList.Value.Error,
+                        ConnectionErrorReason.DeploymentNotFound);
+                }
+
+                return deploymentList.Value;
+            }
+        }
+
+        private static string GetConnectionFailureReason(Connection connection)
+        {
+            return connection.GetConnectionStatusCodeDetailString() ?? "No reason found for connection failure";
```

Comment:
can this actually be null? 
what about the status code?

Reply:
It shouldn't be. I didn't write this though, just moved it I think.